Statsdash/report.py

```diff
@@ -418,6 +400,126 @@ def send_email(self, html):
 
 
 
+class AnalyticsYearSocialReport(Report):
+    def __init__(self, sites, period, recipients, frequency, subject):
+        super(AnalyticsYearSocialReport, self).__init__(sites, period, recipients, frequency, subject)
+        self.sites = sites
+        self.period = period
+        self.recipients = recipients
+        self.frequency = frequency
+        self.subject = subject		 
+        self.data = AnalyticsData(self.sites, self.period, self.frequency)    
+        self.warning_sites = []
+        self.template = self.env.get_template("GA/year_social_report.html")
+        
+        logger.debug("Running analytics top social network report for the year")      
+
+        if len(self.sites) == len(ga_config.TABLES.keys()):
+            self.all_sites = True
+        else:
+            self.all_sites = False
+
+    def get_subject(self):
+        """
+        Return the subject of the email that includes the dates for this period
+        """
+        start_month = date(self.period.start_date.year-1, self.period.start_date.month, 1)
+        dates = start_month.strftime("%B %Y") + " - " + self.period.start_date.strftime('%B %Y')
+        subject = ' '.join([self.subject, dates])
+        
+        return subject
+
+		
+    def check_data_availability(self, override=False):
+        check = self.data.check_available_data()
+        if check["result"]:
+            return True
+        else:
+            logger.debug("Data not available for: %s" % check["site"])
+            if override:
+                self.warning_sites = check["site"]
+                return True
+            else:
+                return False	
+                
+    def get_site(self):
+        if len(self.sites) == 1:
+            return self.sites[0]
+        elif len(self.sites) == len(ga_config.TABLES.keys()):
+            return ga_config.ALL_SITES_NAME
+            
+                        
+    def generate_html(self):
+        
+        today = self.period.end_date
+        
+        start_month = date(today.year-3, today.month, 1)
+        end_month = today + timedelta(days=1) #first of next month so includes this month
+        current = start_month
+        month_stats_range = []
+        while current != end_month:
```

Comment:
A good rule of thumb is to double check while loops' conditions since they have much more potential to be unbounded.  Maybe `current.month is not end_month.month` would be a bit more robust here.  Otherwise, if `utils.add_one_month` somehow changes the day from the 1st of the month, you have potential to run past the target month and loop forever!